parts/k8s/addons/coredns.yaml

```diff
@@ -62,7 +62,10 @@ data:
     import conf.d/Corefile*
     .:53 {
         errors
-        health
+        health {
+          # this should be > readiness probe failure time
+          lameduck 35s
```

Comment:
So the idea here is to fallback to the coredns runtime health check in case readinessProbe fails to restart things after 3 failures separated by 10 seconds?

Reply:
We want to make sure that coredns will not terminate until it is safe for it to do.

The idea here is when coredns gets a SIGTERM, `lameduck` will delay the actual shutdown for 35s. Until then, coredns will continue to service requests. In same time, `readiness` plugin will start failing. We wait 10*3+5 seconds to make sure that this instance is and no longer will be getting any queries.

Reply:
Ah, thanks for the clarification. Just to be super clear, 3 failures, separated by 10 seconds, could be as short as just over *20* seconds. It should never be as long as 30 seconds. So if it helps, we could reduce the lameduck config to 30s.

Reply:
The thing is to delay the eviction of the pod for 35 seconds but trigger health failures right away at the start of the eviction.  This will then, after the health probe failure (3x10) remove the pod from the service (and thus the load balancing for the service) such that shutting it down will not trigger DNS requests to route to the now dead pod.

It was technically a race between not sending any requests to the pod and the pod no longer running.  The old way always ended up with the pod stopping before the requests stopped coming and with this new mechanism this is the inverse now - it always stops sending requests to the pod before the pod stops running.

Reply:
The trick is that there is some time after the health fails that the kubernetes infrastructure removes the pod from the service so the extra 5 seconds is to allow this to fully propagate through the whole cluster.  Remember, some node may have just started to want to do DNS at the same moment that the 30 seconds runs out and thus we could end up routing to the bad pod before the route change got pushed everywhere.

Reply:
Makes sense, lgtm, thanks for this improvement!